File: frontend/src/app/domain/beat.ts
import { Track } from "./track";
import {BPM} from "./bpm";
import { BeatsPerBar } from "./beats-per-bar";
import { SubdivisionsPerBeat } from "./subdivisions-per-beat";
import { NumberOfBar } from "./number-of-bar";

export type Beat = {
  readonly label: string;
  readonly genre: string;
  readonly bpm: BPM;
  readonly beatsPerBar: BeatsPerBar;
  readonly subdivisionsPerBeat: SubdivisionsPerBeat;
  readonly numberOfBar: NumberOfBar;
  readonly tracks: ReadonlyArray<Track>;
}

File: backend/app/domain/Beat.scala
package domain

case class Beat(id: String, label: String, bpm: Int, genre: String, tracks: Seq[Track]) {
  def as[T](implicit f: Beat => T): T = f(this)
}

object Beat {
  implicit def beatMapper: Beat => primary.Beat = (beat: Beat) => primary.Beat(beat.id, beat.label, beat.bpm, beat.genre, beat.tracks.map(x => x.as[primary.Track]))
}